### ISSUE TITLE
Export-CrmSolution failure in 2.8.1

### ISSUE THREAD

**Issue by DingBong** [seanmcne/Microsoft.Xrm.Data.PowerShell#322]
Hi,

I've integrated Microsoft.Xrm.Data.PowerShell into our build pipeline. However, since the update from version 2.8.0 to 2.8.1 I'm receiving an error when I try exporting the solution using `Export-CrmSolution`. Exporting manually works. It produces the following error:

```
2018-09-11T08:48:34.8830701Z ##[command]. 'C:\VsAgent\_work\1\s\BuildScripts\Export-CrmSolution.ps1' -crmUser SVC_CRM_O_SA@domain.local -crmPassword ******** -crmUrl http://server/org -timeOut 30 -outputDir C:\VsAgent\_work\1\a -crmSolution solution -crmOrganizationName orgname -managed $true -unmanaged $true
2018-09-11T08:48:35.2737086Z Logging in as  SVC_CRM_O_SA@domain.local with password  ********
2018-09-11T08:48:35.3205672Z Connecting to CRM
2018-09-11T08:48:35.6018225Z VERBOSE: Verbose output log file: 'C:\Users\SVC_O_BUILD\AppData\Roaming\Microsoft Corporation\Microsoftr Windowsr 
2018-09-11T08:48:35.6018225Z Operating System\10.0.14393.0\Microsoft.Xrm.Tooling.CrmConnector.Powershell-2018-09-11.log'
2018-09-11T08:48:39.3466379Z VERBOSE: Connection Complete
2018-09-11T08:48:39.3466379Z VERBOSE: CRM Connection Timeout set to 30 Minutes
2018-09-11T08:48:39.3466379Z Export CRM Solution orgname to directory  C:\VsAgent\_work\1\a
2018-09-11T08:48:39.3466379Z VERBOSE: Using the supplied single filter of uniquename 'like' solution
2018-09-11T08:48:39.5654938Z VERBOSE: 1 records Found!
2018-09-11T08:48:39.6435464Z VERBOSE: Elapsed Time: 71
2018-09-11T08:48:39.6591490Z VERBOSE: Crm Server responded in:     00:00:00.1302812
2018-09-11T08:48:39.6591490Z VERBOSE: Get-CrmRecordsByFetch Parse: 00:00:00.1154686
2018-09-11T08:48:39.6591490Z VERBOSE: Get-CrmRecordsByFetch Total: 00:00:00.2807292
2018-09-11T08:48:39.8153944Z ##[error]Export-CrmSolution : Solution with name "solution" in CRM Instance: "Org Display Name" not found!
```

I forced the Microsoft.Xrm.Data.PowerShell version back to 2.8.0. That solved the issue.

My configuration:
1. D365 On-Premise - version 1612 (8.2.2.112)
2. AD-Authentication
3. Powershell is executed on build machine in same VLAN, same firewall settings

Below the log file of the broken build.
```
Microsoft.Xrm.Tooling.CrmConnectControl	Information	8	9/11/2018 10:48:35 AM  SetConfigKeyInfo, Key Count = 9
Microsoft.Xrm.Tooling.CrmConnectControl	Information	8	9/11/2018 10:48:35 AM  Using CRM deployment type Prem
Microsoft.Xrm.Tooling.CrmConnectControl	Information	8	9/11/2018 10:48:35 AM  SSL Connection = False
Microsoft.Xrm.Tooling.CrmConnectControl	Information	8	9/11/2018 10:48:35 AM  Discovery URI is = http://server/XRMServices/2011/Discovery.svc
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Information	8	9/11/2018 10:48:35 AM  DiscoverOrganizations - Initializing Discovery Server Object with http://server/XRMServices/2011/Discovery.svc
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/11/2018 10:48:35 AM  DiscoverOrganizations - attempting to connect to CRM server @ http://server/XRMServices/2011/Discovery.svc
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/11/2018 10:48:35 AM  DiscoverOrganizations - created CRM server proxy configuration for http://server/XRMServices/2011/Discovery.svc - duration: 00:00:00.1149858
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/11/2018 10:48:35 AM  DiscoverOrganizations - proxy requiring authentication type : ActiveDirectory 
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/11/2018 10:48:35 AM  DiscoverOrganizations - service proxy created - total create duration: 00:00:00.1197913
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Information	8	9/11/2018 10:48:38 AM  DiscoverOrganizations - Discovery Server Get Orgs Call Complete - Elapsed:00:00:02.9910863
Microsoft.Xrm.Tooling.CrmConnectControl	Information	8	9/11/2018 10:48:38 AM  Found 1 Org(s)
Microsoft.Xrm.Tooling.CrmConnectControl	Verbose	16	9/11/2018 10:48:38 AM  DiscoveryServer indicated organization service location = http://server/orgname/XRMServices/2011/Organization.svc
Microsoft.Xrm.Tooling.CrmConnectControl	Information	8	9/11/2018 10:48:38 AM  Organization Service URI is = http://server/orgname/XRMServices/2011/Organization.svc
Microsoft.Xrm.Tooling.CrmConnectControl	Information	8	9/11/2018 10:48:38 AM  Validating connection to Microsoft Dynamics CRM...
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/11/2018 10:48:38 AM  Initialize CRM connection Started - AuthType: AD
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Information	8	9/11/2018 10:48:38 AM  Process is bypassed.. OrgDetail object was provided
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/11/2018 10:48:38 AM  DiscoveryServer indicated organization service location = http://server/orgname/XRMServices/2011/Organization.svc
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Information	8	9/11/2018 10:48:38 AM  Organization Service URI is = http://server/orgname/XRMServices/2011/Organization.svc
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/11/2018 10:48:38 AM  ConnectAndInitCrmOrgService - Initializing Organization Service Object
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Information	8	9/11/2018 10:48:38 AM  ConnectAndInitCrmOrgService - Requesting connection to Organization with CRM Version: 8.2.2.112
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Information	8	9/11/2018 10:48:38 AM  ConnectAndInitCrmOrgService - Using ISerivceManagement
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/11/2018 10:48:38 AM  ConnectAndInitCrmOrgService - attempting to connect to CRM server @ http://server/orgname/XRMServices/2011/Organization.svc
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/11/2018 10:48:38 AM  ConnectAndInitCrmOrgService - created CRM server proxy configuration for http://server/orgname/XRMServices/2011/Organization.svc - duration: 00:00:00.0769712
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/11/2018 10:48:38 AM  ConnectAndInitCrmOrgService - proxy requiring authentication type : ActiveDirectory 
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/11/2018 10:48:38 AM  ConnectAndInitCrmOrgService - service proxy created - total create duration: 00:00:00.0774748
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Information	8	9/11/2018 10:48:38 AM  ConnectAndInitCrmOrgService - Proxy created, total elapsed time: 00:00:00.0780489
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Information	8	9/11/2018 10:48:38 AM  Beginning Validation of CRM Connection. RequestID: 84905d69-ec76-43c1-b0ce-f2e90904ddb3
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Information	8	9/11/2018 10:48:39 AM  Validation of CRM Connection Complete, total duration: 00:00:00.1408806
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/11/2018 10:48:39 AM  New Batch Manager Created, Max #of Batches:50000, Max #of RequestsPerBatch:5000
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/11/2018 10:48:39 AM  Executed Command - RetrieveMultiple : GetEntityDataByFetchSearch : RequestID=224973e2-ed65-4811-853f-61e3fb6a0b06 : duration=00:00:00.0548979
```

Can we find out what changed?

Thanks!

**Comment by DingBong** [comment]
Some additional information: Updating the CRM solution version, using `Set-CrmRecord`, works fine.

**Comment by seanmcne** [comment]
Got it - the issue is the object type switching from List to PSCustomObject - I've pulled 2.8.1 from the downloads right now and will get this fixed hopefully tomorrow. 

So sorry for the inconvenience! 

![image](https://user-images.githubusercontent.com/2292260/45464142-d0f77380-b6d4-11e8-95cf-f540ec6ff8a7.png)

**Comment by seanmcne** [comment]
Branch updated here: https://github.com/seanmcne/Microsoft.Xrm.Data.PowerShell/blob/seanmcne-281Fix/Microsoft.Xrm.Data.PowerShell/Microsoft.Xrm.Data.PowerShell.psm1 


Please take a few minutes (if you can spare) to test out the proposed changes (you can download the psm1 and replace the one you have in your 2.8.1 directory).  The issue was an implicit cast of a generic last to a generic custom object which lacks a Count() member/function which is why it was resulting in a null value.

**Comment by DingBong** [comment]
Hi Sean,

It's alive 👍😄. I replaced the PSM1 file with the link you provided. The solution exports just like 2.8.0. To test it, I forced `Install-Module` to use 2.8.1 with the modified PSM1. Scripts/Output below:

```
Install-PackageProvider nuget -force
Import-Module Microsoft.Xrm.Data.Powershell -Verbose -RequiredVersion 2.8.1
Install-Module Microsoft.Xrm.Data.PowerShell -Scope CurrentUser -Force -RequiredVersion 2.8.1
```

```
2018-09-13T06:30:16.4489508Z ##[command]. 'C:\VsAgent\_work\1\s\BuildScripts\Export-CrmSolution.ps1' -crmUser SVC_CRM_O_SA@domain.local -crmPassword ******** -crmUrl http://server/orgname -timeOut 30 -outputDir C:\VsAgent\_work\1\a -crmSolution solution -crmOrganizationName orgname -managed $true -unmanaged $true
2018-09-13T06:30:16.8229721Z Logging in as SVC_CRM_O_SA@domain.local with password ********
2018-09-13T06:30:16.8541553Z Connecting to CRM
2018-09-13T06:30:17.1041612Z VERBOSE: Verbose output log file: 'C:\Users\SVC_O_BUILD\AppData\Roaming\Microsoft Corporation\Microsoftr Windowsr 
2018-09-13T06:30:17.1041612Z Operating System\10.0.14393.0\Microsoft.Xrm.Tooling.CrmConnector.Powershell-2018-09-13.log'
2018-09-13T06:30:20.8858804Z VERBOSE: Connection Complete
2018-09-13T06:30:21.0108410Z VERBOSE: CRM Connection Timeout set to 30 Minutes
2018-09-13T06:30:21.0108410Z Export CRM Solution solution to directory C:\VsAgent\_work\1\a
2018-09-13T06:30:21.1046274Z VERBOSE: Using the supplied single filter of uniquename 'like' solution
2018-09-13T06:30:21.3076819Z VERBOSE: 1 records Found!
2018-09-13T06:30:21.3858759Z VERBOSE: Elapsed Time: 77
2018-09-13T06:30:21.4170623Z VERBOSE: Crm Server responded in:     00:00:00.0791824
2018-09-13T06:30:21.4170623Z VERBOSE: Get-CrmRecordsByFetch Parse: 00:00:00.1239021
2018-09-13T06:30:21.4170623Z VERBOSE: Get-CrmRecordsByFetch Total: 00:00:00.2835670
2018-09-13T06:30:21.4327187Z VERBOSE: Solution found with version# 1.90.1.531
2018-09-13T06:30:21.4640086Z VERBOSE: Solution path: C:\VsAgent\_work\1\a\solution_managed.zip
2018-09-13T06:30:21.5108049Z VERBOSE: ExportSolutionRequests may take several minutes to complete execution.
2018-09-13T06:31:03.9990547Z VERBOSE: Using solution file to path: $path
2018-09-13T06:31:04.0146694Z VERBOSE: Successfully wrote file
2018-09-13T06:31:46.9779345Z ExportSolutionResponse                            SolutionPath                          
2018-09-13T06:31:46.9779345Z ----------------------                            ------------                          
2018-09-13T06:31:46.9779345Z Microsoft.Crm.Sdk.Messages.ExportSolutionResponse C:\VsAgent\_work\1\a\solution_managed.zip
2018-09-13T06:31:47.0091769Z ##[section]Finishing: Export <solution> Solution
```

Output log below:
```
Microsoft.Xrm.Tooling.CrmConnectControl	Information	8	9/13/2018 8:31:48 AM  SetConfigKeyInfo, Key Count = 9
Microsoft.Xrm.Tooling.CrmConnectControl	Information	8	9/13/2018 8:31:48 AM  Using CRM deployment type Prem
Microsoft.Xrm.Tooling.CrmConnectControl	Information	8	9/13/2018 8:31:48 AM  SSL Connection = False
Microsoft.Xrm.Tooling.CrmConnectControl	Information	8	9/13/2018 8:31:48 AM  Discovery URI is = http://server/XRMServices/2011/Discovery.svc
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Information	8	9/13/2018 8:31:48 AM  DiscoverOrganizations - Initializing Discovery Server Object with http://server/XRMServices/2011/Discovery.svc
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/13/2018 8:31:48 AM  DiscoverOrganizations - attempting to connect to CRM server @ http://server/XRMServices/2011/Discovery.svc
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/13/2018 8:31:48 AM  DiscoverOrganizations - created CRM server proxy configuration for http://server/XRMServices/2011/Discovery.svc - duration: 00:00:00.1194631
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/13/2018 8:31:48 AM  DiscoverOrganizations - proxy requiring authentication type : ActiveDirectory 
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/13/2018 8:31:48 AM  DiscoverOrganizations - service proxy created - total create duration: 00:00:00.1273919
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Information	8	9/13/2018 8:31:48 AM  DiscoverOrganizations - Discovery Server Get Orgs Call Complete - Elapsed:00:00:00.1442421
Microsoft.Xrm.Tooling.CrmConnectControl	Information	8	9/13/2018 8:31:48 AM  Found 1 Org(s)
Microsoft.Xrm.Tooling.CrmConnectControl	Verbose	16	9/13/2018 8:31:48 AM  DiscoveryServer indicated organization service location = http://server/orgname/XRMServices/2011/Organization.svc
Microsoft.Xrm.Tooling.CrmConnectControl	Information	8	9/13/2018 8:31:48 AM  Organization Service URI is = http://server/orgname/XRMServices/2011/Organization.svc
Microsoft.Xrm.Tooling.CrmConnectControl	Information	8	9/13/2018 8:31:48 AM  Validating connection to Microsoft Dynamics CRM...
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/13/2018 8:31:48 AM  Initialize CRM connection Started - AuthType: AD
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Information	8	9/13/2018 8:31:48 AM  Process is bypassed.. OrgDetail object was provided
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/13/2018 8:31:48 AM  DiscoveryServer indicated organization service location = http://server/orgname/XRMServices/2011/Organization.svc
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Information	8	9/13/2018 8:31:48 AM  Organization Service URI is = http://server/orgname/XRMServices/2011/Organization.svc
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/13/2018 8:31:48 AM  ConnectAndInitCrmOrgService - Initializing Organization Service Object
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Information	8	9/13/2018 8:31:48 AM  ConnectAndInitCrmOrgService - Requesting connection to Organization with CRM Version: 8.2.2.112
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Information	8	9/13/2018 8:31:48 AM  ConnectAndInitCrmOrgService - Using ISerivceManagement
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/13/2018 8:31:48 AM  ConnectAndInitCrmOrgService - attempting to connect to CRM server @ http://server/orgname/XRMServices/2011/Organization.svc
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/13/2018 8:31:48 AM  ConnectAndInitCrmOrgService - created CRM server proxy configuration for http://server/orgname/XRMServices/2011/Organization.svc - duration: 00:00:00.2074382
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/13/2018 8:31:48 AM  ConnectAndInitCrmOrgService - proxy requiring authentication type : ActiveDirectory 
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/13/2018 8:31:48 AM  ConnectAndInitCrmOrgService - service proxy created - total create duration: 00:00:00.2079304
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Information	8	9/13/2018 8:31:48 AM  ConnectAndInitCrmOrgService - Proxy created, total elapsed time: 00:00:00.2086327
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Information	8	9/13/2018 8:31:48 AM  Beginning Validation of CRM Connection. RequestID: fc8bbcb6-7b45-4cb3-9f23-042ac243f6eb
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Information	8	9/13/2018 8:31:48 AM  Validation of CRM Connection Complete, total duration: 00:00:00.1275589
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/13/2018 8:31:48 AM  New Batch Manager Created, Max #of Batches:50000, Max #of RequestsPerBatch:5000
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/13/2018 8:31:49 AM  Executed Command - RetrieveMultiple : GetEntityDataByFetchSearch : RequestID=be9cc97f-da61-4a9a-92b6-9aed9bd10331 : duration=00:00:00.0550140
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/13/2018 8:32:00 AM  Executed Command - ExportSolution : User Defined : RequestID=35c0c43e-9306-4e95-8e77-50a02998a44d : duration=00:00:11.0178357
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/13/2018 8:32:00 AM  Executed Command - RetrieveMultiple : GetEntityDataByFetchSearch : RequestID=ed95fd8f-79fa-4161-b475-94beee858a16 : duration=00:00:00.0150197
Microsoft.Xrm.Tooling.Connector.CrmServiceClient	Verbose	16	9/13/2018 8:32:11 AM  Executed Command - ExportSolution : User Defined : RequestID=210fa551-5311-4f12-b991-880001be05bf : duration=00:00:10.6125092
```

**Comment by seanmcne** [comment]
Phenomenal - thanks for confirming! I'll get this updated and shipped to the gallery as 2.8.2 today.

**Comment by seanmcne** [comment]
The fix is in the gallery as 2.8.1.1 - thanks again for your patience and for reporting this issue!

**Comment by seanmcne** [comment]
👍 Awesome - thanks for confirming!
